feat: new solution deleteDups

truthgoddess opened the pull request:
Hi!

This is a solution that is drawn from my TS solution. 

It is based upon the best time complexity case, but uses getters and setters, and may have some interest to people. If it isn't different enough, awesome.

truthgoddess committed:
feat: new solution deleteDups

jiannapark reviewed (commented):
Nice work. I have two questions, I'm not familiar with the use of underscore (as in this._head), does it have a special meaning to it? And in lines 39, 41 and 44, what does the question mark do?

Also line 73 doesn't seem necessary, as the pointer is reassigned a value right after

truthgoddess committed:
bug: removed unnecessary line

truthgoddess commented:
> Nice work. I have two questions, I'm not familiar with the use of underscore (as in this._head), does it have a special meaning to it? 

Yes. Generally speaking, it is customary to write an _ when it is that that variable is intended to be private (not directly accessed in code, but only used by other functions in the scope of the object). 

> And in lines 39, 41 and 44, what does the question mark do?

The question mark admits to the possibility that something might be undefined, and to return undefined if it is not there. In TypeScript, you have to be very specific about what your return values are, and if it is possible that something is going to be undefined, you have to account for it. In those cases, since the property they are going to be looking for might not be defined, it is used. 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html
 
> Also line 73 doesn't seem necessary, as the pointer is reassigned a value right after

You were right! I've changed that and updated the branch! Thank you for looking so carefully at it!

jiannapark commented:
That's very cool to know, thank you for the explanation! :)
